File: src/main/java/destiny/core/chinese/StemBranch.java
package destiny.core.chinese;

import destiny.tools.ArrayTools;
import org.jetbrains.annotations.NotNull;

import java.io.Serializable;
import java.util.*;

import static destiny.core.chinese.Branch.*;
import static destiny.core.chinese.Stem.*;

/**
 * 中國干支組合表示法，0[甲子] ~ 59[癸亥]
 */
public class StemBranch extends StemBranchOptional implements Comparable<StemBranch> , Serializable {

  public final static StemBranch 甲子 = new StemBranch(甲 , 子);
  public final static StemBranch 乙丑 = new StemBranch(乙 , 丑);
  public final static StemBranch 丙寅 = new StemBranch(丙 , 寅);
  public final static StemBranch 丁卯 = new StemBranch(丁 , 卯);
  public final static StemBranch 戊辰 = new StemBranch(戊 , 辰);
  public final static StemBranch 己巳 = new StemBranch(己 , 巳);
  public final static StemBranch 庚午 = new StemBranch(庚 , 午);
  public final static StemBranch 辛未 = new StemBranch(辛 , 未);
  public final static StemBranch 壬申 = new StemBranch(壬 , 申);
  public final static StemBranch 癸酉 = new StemBranch(癸 , 酉);
  public final static StemBranch 甲戌 = new StemBranch(甲 , 戌);
  public final static StemBranch 乙亥 = new StemBranch(乙 , 亥);
  public final static StemBranch 丙子 = new StemBranch(丙 , 子);
  public final static StemBranch 丁丑 = new StemBranch(丁 , 丑);
  public final static StemBranch 戊寅 = new StemBranch(戊 , 寅);
  public final static StemBranch 己卯 = new StemBranch(己 , 卯);
  public final static StemBranch 庚辰 = new StemBranch(庚 , 辰);
  public final static StemBranch 辛巳 = new StemBranch(辛 , 巳);
  public final static StemBranch 壬午 = new StemBranch(壬 , 午);
  public final static StemBranch 癸未 = new StemBranch(癸 , 未);
  public final static StemBranch 甲申 = new StemBranch(甲 , 申);
  public final static StemBranch 乙酉 = new StemBranch(乙 , 酉);
  public final static StemBranch 丙戌 = new StemBranch(丙 , 戌);
  public final static StemBranch 丁亥 = new StemBranch(丁 , 亥);
  public final static StemBranch 戊子 = new StemBranch(戊 , 子);
  public final static StemBranch 己丑 = new StemBranch(己 , 丑);
  public final static StemBranch 庚寅 = new StemBranch(庚 , 寅);
  public final static StemBranch 辛卯 = new StemBranch(辛 , 卯);
  public final static StemBranch 壬辰 = new StemBranch(壬 , 辰);
  public final static StemBranch 癸巳 = new StemBranch(癸 , 巳);
  public final static StemBranch 甲午 = new StemBranch(甲 , 午);
  public final static StemBranch 乙未 = new StemBranch(乙 , 未);
  public final static StemBranch 丙申 = new StemBranch(丙 , 申);
  public final static StemBranch 丁酉 = new StemBranch(丁 , 酉);
  public final static StemBranch 戊戌 = new StemBranch(戊 , 戌);
  public final static StemBranch 己亥 = new StemBranch(己 , 亥);
  public final static StemBranch 庚子 = new StemBranch(庚 , 子);
  public final static StemBranch 辛丑 = new StemBranch(辛 , 丑);
  public final static StemBranch 壬寅 = new StemBranch(壬 , 寅);
  public final static StemBranch 癸卯 = new StemBranch(癸 , 卯);
  public final static StemBranch 甲辰 = new StemBranch(甲 , 辰);
  public final static StemBranch 乙巳 = new StemBranch(乙 , 巳);
  public final static StemBranch 丙午 = new StemBranch(丙 , 午);
  public final static StemBranch 丁未 = new StemBranch(丁 , 未);
  public final static StemBranch 戊申 = new StemBranch(戊 , 申);
  public final static StemBranch 己酉 = new StemBranch(己 , 酉);
  public final static StemBranch 庚戌 = new StemBranch(庚 , 戌);
  public final static StemBranch 辛亥 = new StemBranch(辛 , 亥);
  public final static StemBranch 壬子 = new StemBranch(壬 , 子);
  public final static StemBranch 癸丑 = new StemBranch(癸 , 丑);
  public final static StemBranch 甲寅 = new StemBranch(甲 , 寅);
  public final static StemBranch 乙卯 = new StemBranch(乙 , 卯);
  public final static StemBranch 丙辰 = new StemBranch(丙 , 辰);
  public final static StemBranch 丁巳 = new StemBranch(丁 , 巳);
  public final static StemBranch 戊午 = new StemBranch(戊 , 午);
  public final static StemBranch 己未 = new StemBranch(己 , 未);
  public final static StemBranch 庚申 = new StemBranch(庚 , 申);
  public final static StemBranch 辛酉 = new StemBranch(辛 , 酉);
  public final static StemBranch 壬戌 = new StemBranch(壬 , 戌);
  public final static StemBranch 癸亥 = new StemBranch(癸 , 亥);

  // 0[甲子] ~ 59[癸亥]
  @NotNull
  private final static StemBranch[] ARRAY = {甲子,乙丑,丙寅,丁卯,戊辰,己巳,庚午,辛未,壬申,癸酉,甲戌,乙亥,丙子,丁丑,戊寅,己卯,庚辰,辛巳,壬午,癸未,甲申,乙酉,丙戌,丁亥,戊子,己丑,庚寅,辛卯,壬辰,癸巳,甲午,乙未,丙申,丁酉,戊戌,己亥,庚子,辛丑,壬寅,癸卯,甲辰,乙巳,丙午,丁未,戊申,己酉,庚戌,辛亥,壬子,癸丑,甲寅,乙卯,丙辰,丁巳,戊午,己未,庚申,辛酉,壬戌,癸亥};

  private final static List<StemBranch> list = Arrays.asList(ARRAY);

  private StemBranch(@NotNull Stem stem, @NotNull Branch branch) {
    super(Optional.of(stem), Optional.of(branch));
  }
  
  /**
   * @param index 0[甲子] ~ 59[癸亥]
   */
  public static StemBranch get(int index) {
    return ArrayTools.INSTANCE.get(ARRAY , index);
  }

  /**
   * 取得下 n 組干支組合
   * n = 0 : 傳回自己
   */
  public StemBranch next(int n) {
    return get(getIndex(this) + n);
  }

  @Override
  public Optional<? extends StemBranchOptional> nextOpt(int n) {
    return Optional.of(next(n));
  }

  /**
   * 取得前 n 組干支組合
   * n = 0 : 傳回自己
   */
  public StemBranch prev(int n) {
    return next(0-n);
  }

  /** 取得下一組干支 , 甲子 傳回 乙丑 */
  public StemBranch getNext() {
    return next(1);
  }

  /** 取得上一組干支 , 甲子 傳回 癸亥 */
  public StemBranch getPrevious() {
    return prev(1);
  }

  @NotNull
  public static StemBranch get(@NotNull Stem stem, @NotNull Branch branch)
  {
    if ( (Stem.getIndex(stem) % 2 )  != (Branch.getIndex(branch) %2 ) )
        throw new RuntimeException("Stem/Branch combination illegal ! " + stem + " cannot be combined with " + branch);

    int sIndex = Stem.getIndex(stem);
    int bIndex = Branch.getIndex(branch);
    switch (sIndex - bIndex) {
      case 0:
      case -10:
        return get(bIndex);
      case 2:
      case -8:
        return get(bIndex + 12);
      case 4:
      case -6:
        return get(bIndex + 24);
      case 6:
      case -4:
        return get(bIndex + 36);
      case 8:
      case -2:
        return get(bIndex + 48);
      default:
        throw new AssertionError("Invalid stem/branch Combination!");
    }
  }

  public static StemBranch get(char stemChar , char branchChar) {
    Optional<Stem> stemOptional = Stem.get(stemChar);
    Optional<Branch> branchOptional = Branch.get(branchChar);

    return stemOptional.map(stem1 -> StemBranch.get(stem1 , branchOptional.orElseThrow(() -> new RuntimeException("Cannot find Branch: " + branchChar))))
      .orElseThrow(() -> new RuntimeException("Cannot get StemBranch("+stemChar+" , "+branchChar+")"));

  }
  
  public static StemBranch get(@NotNull String stemBranch)
  {
    if (stemBranch.length() != 2)
      throw new RuntimeException("The length of " + stemBranch + " must equal to 2 !");
    else
      return get(stemBranch.charAt(0) , stemBranch.charAt(1));
  }
  
  /** 取得干支的差距，例如 "乙丑" 距離 "甲子" 的差距為 "1" , 通常是用於計算「虛歲」 (尚需加一) */
  public int differs(@NotNull StemBranch sb)
  {
    return getIndex(this) - sb.getIndex();
  }

  /**
   * 取得此干支，領先另一組，多少步. 其值一定為正值
   *
   * 「甲子」領先「癸亥」 1
   * 「甲子」領先「乙丑」59
   */
  public int getAheadOf(StemBranch other) {
    int steps = getIndex() - other.getIndex();
    return (steps >=0 ? steps : steps+60);
  }


  /**
   * 0[甲子] ~ 59[癸亥]
   * @param sb 取得某組干支的順序
   * @return 0[甲子] ~ 59[癸亥]
   */
  private static int getIndex(@NotNull StemBranch sb) {
    //return Arrays.binarySearch(ARRAY , sb);
    return list.indexOf(sb);
  }//getIndex()




  /**
   * @return 0[甲子] ~ 59[癸亥]
   */
  public int getIndex()
  {
    return getIndex(this);
  }

  public static Iterator<StemBranch> iterator() {
    return Arrays.stream(ARRAY).iterator();
  }
  
  @NotNull
  public String toString()
  {
    assert (stem != null && branch != null);
    return stem.toString()+branch.toString();
  }

  /**
   * 實作 Comparable 的 compareTo()
   */
  public int compareTo(@NotNull StemBranch o) {
    return (getIndex(this) - getIndex(o));
  }//compareTo()
 
  /**
   * @return stem
   */
  @NotNull
  public Stem getStem()
  {
    assert stem != null;
    return stem;
  }
  
  /**
   * @return branch
   */
  @NotNull
  public Branch getBranch()
  {
    assert branch != null;
    return branch;
  }

  public static List<StemBranch> getList() {
    return Arrays.asList(ARRAY);
  }

  public static Iterable<StemBranch> iterable() {
    return Arrays.asList(ARRAY);
  }

  /** 取得「空亡」的兩個地支 */
  public static Collection<Branch> getEmpties(StemBranch sb) {
    int shift = sb.getStem().getIndex() - sb.getBranch().getIndex();
    switch (shift) {
      case 0 :
        return Arrays.asList(戌 , 亥);
      case 2 :
      case -10 :
        return Arrays.asList(申 , 酉);
      case 4 :
      case -8 :
        return Arrays.asList(午 , 未);
      case 6 :
      case -6 :
        return Arrays.asList(辰 , 巳);
      case 8 :
      case -4 :
        return Arrays.asList(寅 , 卯);
      case 10 :
      case -2:
        return Arrays.asList(子 , 丑);
      default: throw new AssertionError("Cannot find 空亡 from " + sb);
    }
  }

  /** 取得「空亡」的兩個地支 */
  public Collection<Branch> getEmpties() {
    return StemBranch.getEmpties(this);
  }

}

File: src/main/java/destiny/core/chinese/StemBranch.kt
package destiny.core.chinese

import destiny.core.ILoop
import destiny.core.chinese.Branch.*
import destiny.core.chinese.Stem.*
import destiny.core.chinese.StemBranch.甲子
import destiny.core.chinese.StemBranch.癸亥
import destiny.tools.ArrayTools

/**
 * non-nullable && unConstrained
 */
interface IStemBranch : IStemBranchOptional , ILoop<IStemBranch> {
  override val stem: Stem
  override val branch: Branch
}

enum class StemBranchCycle(val sb: StemBranch) {
  甲子(StemBranch.甲子),
  甲寅(StemBranch.甲寅),
  甲辰(StemBranch.甲辰),
  甲午(StemBranch.甲午),
  甲申(StemBranch.甲申),
  甲戌(StemBranch.甲戌)
}

/**
 * 中國干支組合表示法，0[甲子] ~ 59[癸亥]
 */
enum class StemBranch(override val stem: Stem, override val branch: Branch) : IStemBranchOptional, IStemBranch,
  ILoop<IStemBranch>,
  Comparable<StemBranch> {
  甲子(甲, 子),
  乙丑(乙, 丑),
  丙寅(丙, 寅),
  丁卯(丁, 卯),
  戊辰(戊, 辰),
  己巳(己, 巳),
  庚午(庚, 午),
  辛未(辛, 未),
  壬申(壬, 申),
  癸酉(癸, 酉),
  甲戌(甲, 戌),
  乙亥(乙, 亥),
  丙子(丙, 子),
  丁丑(丁, 丑),
  戊寅(戊, 寅),
  己卯(己, 卯),
  庚辰(庚, 辰),
  辛巳(辛, 巳),
  壬午(壬, 午),
  癸未(癸, 未),
  甲申(甲, 申),
  乙酉(乙, 酉),
  丙戌(丙, 戌),
  丁亥(丁, 亥),
  戊子(戊, 子),
  己丑(己, 丑),
  庚寅(庚, 寅),
  辛卯(辛, 卯),
  壬辰(壬, 辰),
  癸巳(癸, 巳),
  甲午(甲, 午),
  乙未(乙, 未),
  丙申(丙, 申),
  丁酉(丁, 酉),
  戊戌(戊, 戌),
  己亥(己, 亥),
  庚子(庚, 子),
  辛丑(辛, 丑),
  壬寅(壬, 寅),
  癸卯(癸, 卯),
  甲辰(甲, 辰),
  乙巳(乙, 巳),
  丙午(丙, 午),
  丁未(丁, 未),
  戊申(戊, 申),
  己酉(己, 酉),
  庚戌(庚, 戌),
  辛亥(辛, 亥),
  壬子(壬, 子),
  癸丑(癸, 丑),
  甲寅(甲, 寅),
  乙卯(乙, 卯),
  丙辰(丙, 辰),
  丁巳(丁, 巳),
  戊午(戊, 午),
  己未(己, 未),
  庚申(庚, 申),
  辛酉(辛, 酉),
  壬戌(壬, 戌),
  癸亥(癸, 亥);

  /** @return 0[甲子] ~ 59[癸亥] */
  val index: Int
    get() = getIndex(this)

  /**
   * 取得下 n 組干支組合
   * n = 0 : 傳回自己
   */
  override fun next(n: Int): StemBranch {
    return get(getIndex(this) + n)
  }

  /**
   * 取得此干支，領先另一組，多少步. 其值一定為正值
   *
   * 「甲子」領先「癸亥」 1
   * 「甲子」領先「乙丑」59
   */
  fun getAheadOf(other: StemBranch): Int {
    val steps = index - other.index
    return if (steps >= 0) steps else steps + 60
  }

  /** 取得干支的差距，例如 "乙丑" 距離 "甲子" 的差距為 "1" , 通常是用於計算「虛歲」 (尚需加一)  */
  fun differs(sb: StemBranch): Int {
    return getIndex(this) - sb.index
  }

  /**
   * 0[甲子] ~ 59[癸亥]
   * @param sb 取得某組干支的順序
   * @return 0[甲子] ~ 59[癸亥]
   */
  private fun getIndex(sb: StemBranch): Int {
    return values().indexOf(sb)
  } //getIndex()

  /** 取得「空亡」的兩個地支  */
  val empties: Collection<Branch>
    get() = getEmpties(this)

  /** 哪一「旬」 */
  val cycle: StemBranchCycle
    get() = getCycle(this)


  override fun toString(): String {
    return stem.toString() + branch.toString()
  }


  companion object {

    /**
     * @param index 0[甲子] ~ 59[癸亥]
     */
    operator fun get(index: Int): StemBranch {
      return ArrayTools[values(), index]
    }

    operator fun get(stem: Stem, branch: Branch): StemBranch {
      if (Stem.getIndex(stem) % 2 != Branch.getIndex(branch) % 2)
        throw RuntimeException("Stem/Branch combination illegal ! $stem cannot be combined with $branch")

      val sIndex = Stem.getIndex(stem)
      val bIndex = Branch.getIndex(branch)
      return when (sIndex - bIndex) {
        0, -10 -> get(bIndex)
        2, -8 -> get(bIndex + 12)
        4, -6 -> get(bIndex + 24)
        6, -4 -> get(bIndex + 36)
        8, -2 -> get(bIndex + 48)
        else -> throw AssertionError("Invalid stem/branch Combination!")
      }
    }

    operator fun get(stemChar: Char, branchChar: Char): StemBranch {

      val stem = Stem[stemChar]
      val branch = Branch[branchChar]
      return if (stem != null && branch != null) {
        StemBranch[stem, branch]
      } else {
        throw RuntimeException("Cannot get StemBranch($stemChar , $branchChar)")
      }
    }

    operator fun get(stemBranch: String): StemBranch {
      return if (stemBranch.length != 2)
        throw RuntimeException("The length of $stemBranch must equal to 2 !")
      else
        get(stemBranch[0], stemBranch[1])
    }

    /** 強制將 [IStemBranchOptional] downcast 成 [StemBranch] , 可能會拋出 NPE ! */
    operator fun get(sbo: IStemBranchOptional): StemBranch {
      return get(sbo.stem!!, sbo.branch!!)
    }


    /**
     * 0[甲子] ~ 59[癸亥]
     * @param sb 取得某組干支的順序
     * @return 0[甲子] ~ 59[癸亥]
     */
    private fun getIndex(sb: StemBranch): Int {
      return values().indexOf(sb)
    } //getIndex()

    operator fun iterator(): Iterator<StemBranch> {
      return values().iterator()
    }

    fun iterable(): Iterable<StemBranch> {
      return listOf(*values())
    }

    fun getCycle(sb: StemBranch): StemBranchCycle {
      return when (sb.stem.index - sb.branch.index) {
        0 -> StemBranchCycle.甲子
        2, -10 -> StemBranchCycle.甲戌
        4, -8 -> StemBranchCycle.甲申
        6, -6 -> StemBranchCycle.甲午
        8, -4 -> StemBranchCycle.甲辰
        10, -2 -> StemBranchCycle.甲寅
        else -> throw AssertionError("Not valid $sb")
      }
    }

    /** 取得「空亡」的兩個地支  */
    fun getEmpties(sb: StemBranch): Collection<Branch> {
      return when (sb.cycle) {
        StemBranchCycle.甲子 -> listOf(戌, 亥)
        StemBranchCycle.甲戌 -> listOf(申, 酉)
        StemBranchCycle.甲申 -> listOf(午, 未)
        StemBranchCycle.甲午 -> listOf(辰, 巳)
        StemBranchCycle.甲辰 -> listOf(寅, 卯)
        StemBranchCycle.甲寅 -> listOf(子, 丑)
      }
    }
  } // companion
}